### DISCUSSION TITLE
RTK Query equivalent for Zustand

### DISCUSSION THREAD

**Discussion by renancleyson-dev** [pmndrs/zustand#1246] [General]
The only third-party library that is missing to me on Zustand for now... The server state can quickly become just boilerplate code.

**Comment by dai-shi** [comment]
Hi, I'm curious how RTK Query equivalent for Zustand looks like. Can you show some small examples (code snippets) to illustrate its usage?
In the end, I think this is the field for 3rd-party libraries.

**Reply by renancleyson-dev** [reply]
Sorry for the long delay, tough September.

I think we need to keep at least the store's structure unopinionated, but the only way I can think of is using immer and doing something like this:
``` javascript
set((state) => {
  getSlice(state).data = data;
  getSlice(state).isLoading = false;
});
```
Now it's possible to create async actions that you can put anywhere on the store and the state can be anywhere else and it will still be updated if you give the `getSlice` function. I don't think it needs to be identical to RTK Query, what I like most is to have my API definition in one place instead of how `react-query` does.

So I was thinking of something like this:
``` javascript
// service/notes.js
import createAPI from 'zustand-query';

export const createEndpoints = (set) => {
  const createEndpoint = createAPI({
    set, // add zustand set here
    getSlice: (state) => state.notes
    prefix: 'notes', // Something for redux devtools, then the action would look like notes/success or notes/request
  });

  // finally define endpoints
  return {
    fetchNotes: createEndpoint(async (page = 1) => {
      const response = await fetch('/notes');
      return response.json();
    })
  }
};

// store.js
import { createEndpoints } from 'services/notes';

// Some zustand store with flux pattern
const useStore = createStore((set) => {
  const endpoints = createEndpoints(set);

  return {
    notes: endpoints.fetchNotes.initialState, // { isLoading, isFetching, data, ... }
    fetchNotes: endpoints.fetchNotes.action, // async action
  };
});

// App.js
const App = () => {
  const { isLoading, data } = useStore((state) => state.notes);
  const fetchNotes = useStore((state) => state.fetchNotes);

  useEffect(() => {
    fetchNotes();
  }, [fetchNotes]);

  if (isLoading) {
     return <div>loading...</div>
  }

  return (
    <div>
      {data.map(note => <div>{note.name}</div>)}
    </div>
  );
};
```

This is what I was thinking of, not sure how caching would work, but let's worry about the usage for now.

**Reply by dai-shi** [reply]
> let's worry about the usage for now.

Okay. My understanding is it's a combination of namespacing/prefixing and loading state tracking.
I would personally use bare zustand store, but yeah such helpers would be possible.
This should be a field for third party libraries. (like you imply with `from 'zustand-query'`)
